File: robot/shared/Shape/FieldObject.cpp
#include "FieldObject.h"

FieldObject::FieldObject(const Vector2D & position, float radius,
                         int fillColor,
                         int strokeColor, float strokeThickness)
  : Circle(position, radius,
           fillColor,
           strokeColor, strokeThickness),
    _havePosition(false),
    _haveAngle(false),
    positionColor(0x00000000),
    angleColor(0x00000000),
    positionDeviation(0),
    angleDeviation(0),
    angle(0) {
  // We need to set type here, since FieldObject is a subclass of Circle
  setType(Shape::FieldObject);
}

FieldObject::FieldObject(const Vector2D & position, float radius,
                         int fillColor,
                         int strokeColor, float strokeThickness,
                         int _positionColor, float _positionDeviation)
  : Circle(position, radius,
           fillColor,
           strokeColor, strokeThickness),
    _havePosition(true),
    _haveAngle(false),
    positionColor(_positionColor),
    angleColor(0x00000000),
    positionDeviation(_positionDeviation),
    angleDeviation(0),
    angle(0) {
  // We need to set type here, since FieldObject is a subclass of Circle
  setType(Shape::FieldObject);
}

FieldObject::FieldObject(const Vector2D & position, float radius,
                         int fillColor,
                         int strokeColor, float strokeThickness,
                         int _angleColor, float _angle, float _angleDeviation)
  : Circle(position, radius,
           fillColor,
           strokeColor, strokeThickness),
    _havePosition(false),
    _haveAngle(true),
    positionColor(0x00000000),
    angleColor(_angleColor),
    positionDeviation(0),
    angleDeviation(_angleDeviation),
    angle(_angle) {
  // We need to set type here, since FieldObject is a subclass of Circle
  setType(Shape::FieldObject);
}

FieldObject::FieldObject(const Vector2D & position, float radius,
                         int fillColor,
                         int strokeColor, float strokeThickness,
                         int _positionColor, float _positionDeviation,
                         int _angleColor, float _angle, float _angleDeviation)
  : Circle(position, radius,
           fillColor,
           strokeColor, strokeThickness),
    _havePosition(true),
    _haveAngle(true),
    positionColor(_positionColor),
    angleColor(_angleColor),
    positionDeviation(_positionDeviation),
    angleDeviation(_angleDeviation),
    angle(_angle) {
  // We need to set type here, since FieldObject is a subclass of Circle
  setType(Shape::FieldObject);
}

FieldObject::~FieldObject() {
}

bool FieldObject::havePosition() const {
  return _havePosition;
}

bool FieldObject::haveAngle() const {
  return _haveAngle;
}

int FieldObject::getPositionColor() const {
  return positionColor;
}

int FieldObject::getAngleColor() const {
  return angleColor;
}

float FieldObject::getPositionDeviation() const {
  return positionDeviation;
}

float FieldObject::getAngle() const {
  return angle;
}

float FieldObject::getAngleDeviation() const {
  return angleDeviation;
}

File: tools/Visualizer/src/shape/FieldObject.java
package shape;

public class FieldObject extends Circle {

  public FieldObject(float x, float y, float radius,
                     int fillColor,
                     int strokeColor, float strokeThickness) {
    super(x, y, radius, fillColor, strokeColor, strokeThickness);
    
    this.havePosition = false;
    this.haveAngle    = false;
  }

  public FieldObject(float x, float y, float radius,
                     int fillColor,
                     int strokeColor, float strokeThickness,
                     int positionColor, float positionDeviation) {
    super(x, y, radius, fillColor, strokeColor, strokeThickness);
    
    this.havePosition = true;
    this.haveAngle    = false;
    this.positionColor     = fixAlpha(positionColor);
    this.positionDeviation = positionDeviation;
  }
  
  public FieldObject(float x, float y, float radius,
                     int fillColor,
                     int strokeColor, float strokeThickness,
                     int angleColor, float angle, float angleDeviation) {
    super(x, y, radius, fillColor, strokeColor, strokeThickness);
    
    this.havePosition = false;
    this.haveAngle    = true;
    this.angleColor     = fixAlpha(angleColor);
    this.angle          = angle;
    this.angleDeviation = angleDeviation;
  }

  public FieldObject(float x, float y, float radius,
                     int fillColor,
                     int strokeColor, float strokeThickness,
                     int positionColor, float positionDeviation,
                     int angleColor, float angle, float angleDeviation) {
    super(x, y, radius, fillColor, strokeColor, strokeThickness);
    
    this.havePosition = true;
    this.haveAngle    = true;
    this.positionColor     = fixAlpha(positionColor);
    this.positionDeviation = positionDeviation;
    this.angleColor     = fixAlpha(angleColor);
    this.angle          = angle;
    this.angleDeviation = angleDeviation;
  }
  
  public boolean havePosition() {
    return havePosition && (positionDeviation > 0);
  }
  
  public boolean haveAngle() {
    return haveAngle;
  }
  
  public int getPositionColor() {
    return positionColor;
  }
  
  public int getAngleColor() {
    return angleColor;
  }
  
  public float getPositionDeviation() {
    return positionDeviation;
  }
  
  public float getAngleDeviation() {
    return angleDeviation;
  }
  
  public float getAngle() {
    return angle;
  }
  
  private int     positionColor,     angleColor;
  private boolean havePosition,      haveAngle;
  private float   positionDeviation, angleDeviation;
  private float   angle;
}
